plan/aggregation_eliminate.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// aggEliminator tries to eliminate max/min aggregate function.
+// For SQL like `select max(id) from t;`, we could optimize it to `select max(id) from (select id from t order by id desc limit 1 where id is not null) t;`.
+// For SQL like `select min(id) from t;`, we could optimize it to `select max(id) from (select id from t order by id limit 1 where id is not null) t;`.
+type aggEliminator struct {
+	ctx context.Context
+}
+
+func (a *aggEliminator) optimize(p LogicalPlan, ctx context.Context) (LogicalPlan, error) {
+	a.ctx = ctx
+	a.eliminateAgg(p)
+	return p, nil
+}
+
+// Try to convert max/min to Limit+Sort operators.
+func (a *aggEliminator) eliminateAgg(p LogicalPlan) {
+	// We don't need to guarantee that the child of it is a data source. This transformation won't be worse than previous.
+	if agg, ok := p.(*LogicalAggregation); ok {
+		// We only consider case with single max/min function.
+		if len(agg.AggFuncs) != 1 || len(agg.GroupByItems) != 0 {
+			return
+		}
+		f := agg.AggFuncs[0]
+		if f.GetName() != ast.AggFuncMax && f.GetName() != ast.AggFuncMin {
+			return
+		}
+		// e.g. select min(-a) from t. We cannot rewrite it easily.
+		if _, ok := f.GetArgs()[0].(*expression.Column); !ok {
+			return
+		}
+
+		child := p.Children()[0]
+		// If it can be NULL, we need to filter NULL out first.
```

Comment:
Could you add tests covers the cases that there is null in the table and the table source result is empty.

Reply:
added